src/Profiling.cpp

```diff
@@ -20,23 +24,154 @@ class InjectProfiling : public IRMutator {
+            }
+        }
+
+        // Check that the allocation is not scalar (if it were scalar
+        // it would have constant size).
+        internal_assert(extents.size() > 0);
+
+        on_stack = false;
+        Expr size = extents[0];
+        for (size_t i = 1; i < extents.size(); i++) {
+            size *= extents[i];
+        }
+        size = simplify(Select::make(condition, size * type.bytes(), 0));
+        return size;
+    }
+
+    void visit(const Allocate *op) {
+        int idx = get_func_id(op->name);
+
+        vector<Expr> new_extents;
+        bool all_extents_unmodified = true;
+        for (size_t i = 0; i < op->extents.size(); i++) {
+            new_extents.push_back(mutate(op->extents[i]));
+            all_extents_unmodified &= new_extents[i].same_as(op->extents[i]);
+        }
+        Expr condition = mutate(op->condition);
+
+        bool on_stack;
+        Expr size = compute_allocation_size(new_extents, condition, op->type, op->name, on_stack);
+        func_alloc_sizes.push(op->name, {on_stack, size});
+
+        if (!is_zero(size) && on_stack) {
+            const int64_t *int_size = as_const_int(size);
+            internal_assert(int_size != NULL); // Stack size is always a const int
+            func_stack_current[idx] += *int_size;
+            func_stack_peak[idx] = std::max(func_stack_peak[idx], func_stack_current[idx]);
+            debug(1) << "  Allocation on stack: " << op->name << "(" << size << ") in pipeline " << pipeline_name
+                     << "; current: " << func_stack_current[idx] << "; peak: " << func_stack_peak[idx] << "\n";
+        }
+
+        Stmt body = mutate(op->body);
+        Expr new_expr;
+        if (op->new_expr.defined()) {
+            new_expr = mutate(op->new_expr);
+        }
+        if (all_extents_unmodified &&
+            body.same_as(op->body) &&
+            condition.same_as(op->condition) &&
+            new_expr.same_as(op->new_expr)) {
+            stmt = op;
+        } else {
+            stmt = Allocate::make(op->name, op->type, new_extents, condition, body, new_expr, op->free_function);
+        }
+
+        //debug(0) << stmt << "\n\n";
+
+        if (!is_zero(size) && !on_stack) {
+            Expr profiler_pipeline_state = Variable::make(Handle(), "profiler_pipeline_state");
+            debug(1) << "  Allocation on heap: " << op->name << "(" << size << ") in pipeline " << pipeline_name << "\n";
+            Expr set_task = Call::make(Int(32), "halide_profiler_memory_allocate",
+                                       {profiler_pipeline_state, idx, size}, Call::Extern);
+            stmt = Block::make(Evaluate::make(set_task), stmt);
+        }
+    }
+
+    void visit(const Free *op) {
+        int idx = get_func_id(op->name);
+
+        AllocSize alloc = func_alloc_sizes.get(op->name);
+        func_alloc_sizes.pop(op->name);
+
+        IRMutator::visit(op);
+
+        if (!is_zero(alloc.size)) {
```

Comment:
In what circumstances is it zero? It's not clear to me why it would ever be zero, so it's worth adding a comment here.